### PR TITLE
Add feeds cog to handle content notifications

### LINKED ISSUE #23: YouTube content notifications
- Announce Carberra videos with @Videos tag in videos channel.
- Maybe support other channels too in sepearate channel, with separate message.

_Originally posted by @Bandarusuhas143 in https://github.com/Carberra/Carberretta/issues/22#issuecomment-700065873_

### LINKED ISSUE #68: Twitch content notifications
Announce Twitch streams as they go live. This will replace KitRobit's in this regard.

### PR DESCRIPTION
**Pull request by Jonxslays**
## Description
Adds a new feeds cog for sending channel notifications about new content uploads.
I did pretty thorough testing, please let me know if there is anything that needs to be fine tuned.
Support for other content creators, with more command functionality in V2?

- Closes #68
- Closes #23 (Mostly)
- Complies with changes outlined in #76

### Dependency additions:
    feedparser
xml is a good alternative to the steep quota cost of searching for video id's using the youtube api.

### .env additions:
    TWITCH_CLIENT_ID
    TWITCH_CLIENT_SECRET

### Database additions:
    "feeds" table
    "premieres" table

## Features
- Distinguishes between videos and vods via "#VOD" being in a vods description.
- Distinguishes between videos and premieres via youtube api attributes.
- Twitch streams will have 1 message that is edited when the stream ends.
- Youtube premieres will have 2 messages:
    - Initial message that a premiere has been scheduled.
    - Secondary message signaling the premiere has begun.

- Add "feed" command group for on demand notifications:
    - "video" subcommand -> Checks for videos and premieres.
    - "vod" subcommand -> Checks for vods.
    - "stream" subcommand -> Checks for Twitch streams.

- Currently runs 3 jobs every 3 minutes, to automate notifications.
- Allows for 3 separate notification roles: videos, vods, streams.
- Data persistence through downtime via the database.

## Quirks
- If the bot is offline for an extended period and multiple videos/vods/premieres are posted, only the newest of each will be announced.
- The youtube xml feed caches causing delays of 5-15 minutes sometimes before videos/vods are announced.
    - I added a different string to the end of each feed call in an attempt to reduce the impact of this.
    - This still allows for more flexibility than the youtube api quota.
    - Twitch streams are unaffected by this behavior.

## Resolves
- [x] Replace KitRobit
- [x] Announce Twitch streams as they go live.
- [x] Announce Carberra videos with @videos tag in videos channel.
- [ ] Maybe support other channels too in separate channel, with separate message.

### CODE REVIEW

File: carberretta/bot/cogs/feeds.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+            elif "#VOD" in item.summary:
+                # This is a vod we havent announced
+
+                await self.videos_channel.send(
+                    f"Hey {self.vods_role.mention}, a new VOD just went live!\nCatch up on anything you missed from the last stream!",
+                    embed=discord.Embed.from_dict(
+                        {
+                            "title": item.title,
+                            "description": desc if len(desc := item.summary) <= 500 else f"{desc[:500]}...",
```
Comment (parafoxia) [inline carberretta/bot/cogs/feeds.py:106]: I actually found out the other day you can use a `textwrap.shorten(...)` call to do this, and it doesn't cut off mid word. It's a builtin library, so have a look. I think it'd be something like `textwrap.shorten(item.summary, 500, placeholder="...")`.

Don't forget to change it below as well!

File: carberretta/bot/cogs/feeds.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+            self.bot.scheduler.add_job(self.get_new_videos, CronTrigger(minute="*/3", second=0))
+            self.bot.scheduler.add_job(self.get_new_vods, CronTrigger(minute="*/3", second=20))
+            self.bot.scheduler.add_job(self.get_new_streams, CronTrigger(minute="*/3", second=40))
+
+            self.bot.ready.up(self)
+
+    async def get_new_vods(self) -> str:
+        current_vod = await self.bot.db.field("SELECT ContentValue FROM feeds WHERE ContentType = ?", "vod")
+
+        for item in data if (data := await self.call_feed()) else []:
```
Comment (parafoxia) [inline carberretta/bot/cogs/feeds.py:89]: You could probably make `call_feed()` return `[]` if the request failed. I didn't actually know this was valid syntax though -- might come in handy later.

File: carberretta/bot/cogs/feeds.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+                        }
+                    ),
+                )
+
+                return data["title"], True
+
+    @commands.group(name="feed", invoke_without_command=True)
+    @commands.is_owner()
+    async def feed_group(self, ctx: commands.Context) -> None:
+        pprint(await self.call_feed())
```
Comment (parafoxia) [inline carberretta/bot/cogs/feeds.py:309]: `pprint`?

File: carberretta/bot/cogs/feeds.py [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+    @feed_group.command(name="video")
+    @commands.is_owner()
+    async def feed_video_command(self, ctx: commands.Context) -> None:
+        last_video = await self.get_new_videos()
+        await ctx.send(f"Announced video: {last_video}" if last_video else "No new videos")
```
Comment (parafoxia) [inline carberretta/bot/cogs/feeds.py:315]: I'm OCD -- add full stops lmao

**Review by parafoxia — CHANGES_REQUESTED** [review]
Just a few small ones.

I also need to add the new IDs before merging this.

File: carberretta/bot/cogs/feeds.py (hunk 1 above)
Comment (Jonxslays) [reply carberretta/bot/cogs/feeds.py:106]: scrapped due to textwrap.shorten() stripping all whitespace.

File: carberretta/bot/cogs/feeds.py [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+
+
+                if not upcoming and duration != "P0D":
+                    # We have not seen this premiere before
+
+                    if live_content == "upcoming" and not announced:
+                        # This premiere is upcoming and not live
+
+                        await self.videos_channel.send(
+                            f"Hey {self.videos_role.mention},\nA new premiere is scheduled for {chron.long_date_and_time(scheduled_time)} UTC!\nHope to see you there!",
```
Comment (parafoxia) [inline carberretta/bot/cogs/feeds.py:192]:
```suggestion
                            f"Hey {self.videos_role.mention}, a new premiere is scheduled for {chron.long_date_and_time(scheduled_time)} UTC! Hope to see you there!",
```

File: carberretta/bot/cogs/feeds.py [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+
+            if current_vod == item.yt_videoid:
+                # We announced this vod already
+                return
+
+            elif "#VOD" in item.summary:
+                # This is a vod we havent announced
+
+                await self.videos_channel.send(
+                    f"Hey {self.vods_role.mention}, a new VOD just went live!\nCatch up on anything you missed from the last stream!",
```
Comment (parafoxia) [inline carberretta/bot/cogs/feeds.py:104]:
```suggestion
                    f"Hey {self.vods_role.mention}, a new VOD just went live! Catch up on anything you missed from the last stream!",
```

File: carberretta/bot/cogs/feeds.py [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+                        )
+
+                        await self.bot.db.commit()
+                        return item.yt_videoid, False
+
+                    elif live_content =="live" and not upcoming and not announced:
+                        # The premiere was never upcoming is now live
+
+                        await self.videos_channel.send(
+                            f"Hey {self.videos_role.mention},\nA new premiere started on {chron.long_date_and_time(scheduled_time)} UTC!\nCome and join us!",
```
Comment (parafoxia) [inline carberretta/bot/cogs/feeds.py:217]:
```suggestion
                            f"Hey {self.videos_role.mention}, a new premiere started on {chron.long_date_and_time(scheduled_time)} UTC! Come and join us!",
```

File: carberretta/bot/cogs/feeds.py [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+                        )
+
+                        await self.bot.db.commit()
+                        return item.yt_videoid, True
+
+                elif not announced:
+                    # A premiere was upcoming, and is now live
+
+                    await self.videos_channel.send(
+                        f"Hey {self.videos_role.mention},\nA new premiere started on {chron.long_date_and_time(scheduled_time)} UTC!\nCome and join us!",
```
Comment (parafoxia) [inline carberretta/bot/cogs/feeds.py:242]:
```suggestion
                        f"Hey {self.videos_role.mention}, a new premiere started on {chron.long_date_and_time(scheduled_time)} UTC! Come and join us!",
```

**Review by parafoxia — CHANGES_REQUESTED** [review]
Finishing touches (:

**Review by parafoxia — APPROVED** [review]

**Pull request merged**